File: packages/ui/dialog.ts
/**
 * Controlled in-window modal dialogs. The Rust core mounts the portal only while the dialog is
 * open and owns focus containment, restoration, dismissal, and the declared transitions.
 */

import { NativePart, type NativeNode, type QuickGuiEvent } from "@quickgui/native";
import { EVENT_CLICK, EVENT_COMPONENT_CHANGE, EVENT_DISMISS } from "@quickgui/native/native-tree";

import {
  CODE_DISMISS_ON_ESCAPE,
  CODE_DISMISS_ON_POINTER_OUTSIDE,
  CODE_ENTER_DURATION,
  CODE_EXIT_DURATION,
  CODE_OPEN,
  CODE_VARIANT,
} from "./generated.ts";
import { componentChangeFromEvent } from "./index.ts";
import {
  createButtonPart,
  createComponentScope,
  createPartContext,
  createViewPart,
  finishPart,
  forwardClick,
  requireContext,
  setPart,
  type PartProps,
} from "./parts.ts";
import { createRenderEffect, signal, type Accessor } from "./reactive.ts";
import { fragment, setExplicitBool, setListener, setMilliseconds, setString } from "./runtime.ts";

export type DialogOpenChangeReason = "trigger-press" | "close-press" | "dismiss";

export interface DialogOpenChangeDetails {
  reason: DialogOpenChangeReason;
  event: QuickGuiEvent;
}

export interface DialogRootProps {
  children?: () => NativeNode;
  open?: Accessor<boolean>;
  defaultOpen?: boolean;
  onOpenChange?: (open: boolean, details: DialogOpenChangeDetails) => void;
  /** The core held the surface mounted for the declared exit transition and finished it. */
  onOpenChangeComplete?: (open: boolean, event: QuickGuiEvent) => void;
  dismissOnEscape?: boolean;
  /** Dismiss on a backdrop press. Defaults to `true` for dialogs and `false` for alert dialogs. */
  dismissOnBackdrop?: boolean;
  /** Enter transition the core waits for, in milliseconds. */
  enterDuration?: number;
  /** Exit transition the core keeps the surface mounted for, in milliseconds. */
  exitDuration?: number;
}

export interface DialogPopupProps extends PartProps {
  onDismiss?: (event: QuickGuiEvent) => void;
}

type DialogVariant = "dialog" | "alertdialog";

class DialogState {
  readonly scope: string;
  readonly variant: DialogVariant;
  readonly props: DialogRootProps;
  readonly _open = signal<boolean>(false);

  constructor(variant: DialogVariant, props: DialogRootProps) {
    this.scope = createComponentScope(variant === "alertdialog" ? "qg-alert-dialog" : "qg-dialog");
    this.variant = variant;
    this.props = props;
    this._open.write(props.defaultOpen ?? false);
  }

  open(): boolean {
    const controlled = this.props.open;
    return controlled !== undefined ? controlled() : this._open.read();
  }

  dismissOnEscape(): boolean {
    return this.props.dismissOnEscape ?? true;
  }

  dismissOnBackdrop(): boolean {
    return this.props.dismissOnBackdrop ?? this.variant !== "alertdialog";
  }

  change(next: boolean, reason: DialogOpenChangeReason, event: QuickGuiEvent): void {
    if (this.props.open === undefined) this._open.write(next);
    const onOpenChange = this.props.onOpenChange;
    if (onOpenChange !== undefined) onOpenChange(next, { reason, event });
  }

  complete(next: boolean, event: QuickGuiEvent): void {
    const onOpenChangeComplete = this.props.onOpenChangeComplete;
    if (onOpenChangeComplete !== undefined) onOpenChangeComplete(next, event);
  }

  applyTo(node: NativeNode, part: string): void {
    setPart(node, part, this.scope, undefined);
    setString(node, CODE_VARIANT, this.variant);
    createRenderEffect(() => {
      setExplicitBool(node, CODE_OPEN, this.open());
    });
  }
}

const DialogContext = createPartContext<DialogState>();

function requireDialog(part: string): DialogState {
  return requireContext(DialogContext, part, "Dialog.Root");
}

function createDialogRoot(variant: DialogVariant, props: DialogRootProps): NativeNode {
  const state = new DialogState(variant, props);
  return DialogContext.provide(state, () => {
    const children = props.children;
    return children === undefined ? fragment([]) : children();
  });
}

function createDialogPart(part: string, props: PartProps, name: string): NativeNode {
  const state = requireDialog(name);
  const node = createViewPart(props);
  state.applyTo(node, part);
  return finishPart(node, props);
}

function createTrigger(props: PartProps): NativeNode {
  const state = requireDialog("Dialog.Trigger");
  const node = createButtonPart(props);
  state.applyTo(node, NativePart.DialogTrigger);
  setListener(
    node,
    EVENT_CLICK,
    forwardClick(props.onClick, (event: QuickGuiEvent): void => {
      state.change(true, "trigger-press", event);
    }),
  );
  return finishPart(node, props);
}

/**
 * Viewport portal, focus trap, and focus-restoration boundary. The core mounts it only while the
 * dialog is open, so a closed dialog contributes no overlay, layout, paint, input, or
 * accessibility node.
 */
function createPortal(props: PartProps): NativeNode {
  const state = requireDialog("Dialog.Portal");
  const node = createViewPart(props);
  state.applyTo(node, NativePart.Dialog);
  setMilliseconds(node, CODE_ENTER_DURATION, state.props.enterDuration);
  setMilliseconds(node, CODE_EXIT_DURATION, state.props.exitDuration);
  setListener(node, EVENT_COMPONENT_CHANGE, (event: QuickGuiEvent): void => {
    const details = componentChangeFromEvent(event);
    if (details === undefined) return;
    const complete = details.openChangeComplete;
    if (complete !== undefined) state.complete(complete, event);
  });
  return finishPart(node, props);
}

/** Modal surface. Escape and backdrop dismissal are decided ahead of time by the core. */
function createPopup(props: DialogPopupProps): NativeNode {
  const state = requireDialog("Dialog.Popup");
  const node = createViewPart(props);
  state.applyTo(node, NativePart.DialogPopup);
  setExplicitBool(node, CODE_DISMISS_ON_ESCAPE, state.dismissOnEscape());
  setExplicitBool(node, CODE_DISMISS_ON_POINTER_OUTSIDE, state.dismissOnBackdrop());
  const onDismiss = props.onDismiss;
  setListener(node, EVENT_DISMISS, (event: QuickGuiEvent): void => {
    if (onDismiss !== undefined) onDismiss(event);
    if (!event.defaultPrevented) state.change(false, "dismiss", event);
  });
  return finishPart(node, props);
}

/** Close control. Its accessible name comes from `ariaLabel`, defaulting to `Close`. */
function createClose(props: PartProps): NativeNode {
  const state = requireDialog("Dialog.Close");
  const node = createButtonPart(props);
  state.applyTo(node, NativePart.DialogClose);
  setListener(
    node,
    EVENT_CLICK,
    forwardClick(props.onClick, (event: QuickGuiEvent): void => {
      state.change(false, "close-press", event);
    }),
  );
  return finishPart(node, props);
}

/** Compound parts for a controlled in-window modal dialog. */
export class Dialog {
  /** Logical root; it creates no native element. */
  static Root(props: DialogRootProps): NativeNode {
    return createDialogRoot("dialog", props);
  }

  /** Trigger button carrying the core's dialog popover and expanded accessibility state. */
  static Trigger(props: PartProps): NativeNode {
    return createTrigger(props);
  }

  static Portal(props: PartProps): NativeNode {
    return createPortal(props);
  }

  /** Scrollable dialog body; a long dialog scrolls inside the popup rather than past the window. */
  static Viewport(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogViewport, props, "Dialog.Viewport");
  }

  /** Application-owned backdrop filling the portal. */
  static Backdrop(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogBackdrop, props, "Dialog.Backdrop");
  }

  static Popup(props: DialogPopupProps): NativeNode {
    return createPopup(props);
  }

  /** Visible dialog title used as the popup's accessible name. */
  static Title(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogTitle, props, "Dialog.Title");
  }

  /** Visible dialog description used as the popup's accessible description. */
  static Description(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogDescription, props, "Dialog.Description");
  }

  static Close(props: PartProps): NativeNode {
    return createClose(props);
  }
}

/** Compound parts for a consequential alert dialog whose backdrop does not dismiss by default. */
export class AlertDialog {
  static Root(props: DialogRootProps): NativeNode {
    return createDialogRoot("alertdialog", props);
  }

  static Trigger(props: PartProps): NativeNode {
    return createTrigger(props);
  }

  static Portal(props: PartProps): NativeNode {
    return createPortal(props);
  }

  static Viewport(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogViewport, props, "AlertDialog.Viewport");
  }

  static Backdrop(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogBackdrop, props, "AlertDialog.Backdrop");
  }

  static Popup(props: DialogPopupProps): NativeNode {
    return createPopup(props);
  }

  static Title(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogTitle, props, "AlertDialog.Title");
  }

  static Description(props: PartProps): NativeNode {
    return createDialogPart(NativePart.DialogDescription, props, "AlertDialog.Description");
  }

  static Close(props: PartProps): NativeNode {
    return createClose(props);
  }
}

File: go/ui/dialog.go
package ui

import (
	"github.com/egoist/quickgui/go/native"
	"github.com/egoist/quickgui/go/protocol"
	"github.com/egoist/quickgui/go/reactive"
)

// DialogOpenChangeReason is why a dialog opened or closed.
type DialogOpenChangeReason string

const (
	DialogTriggerPress DialogOpenChangeReason = "trigger-press"
	DialogClosePress   DialogOpenChangeReason = "close-press"
	DialogDismiss      DialogOpenChangeReason = "dismiss"
)

type DialogOpenChangeDetails struct {
	Reason DialogOpenChangeReason
	Event  *native.Event
}

type DialogRootProps struct {
	Children             Component
	Open                 func() bool
	DefaultOpen          bool
	OnOpenChange         func(bool, DialogOpenChangeDetails)
	OnOpenChangeComplete func(bool, *native.Event)
	DismissOnEscape      *bool
	DismissOnBackdrop    *bool
	EnterDuration        any
	ExitDuration         any
}

type DialogPopupProps struct {
	PartProps
	OnDismiss func(*native.Event)
}

type dialogVariant string

const (
	dialogVariantDialog      dialogVariant = "dialog"
	dialogVariantAlertDialog dialogVariant = "alertdialog"
)

type dialogState struct {
	scope   string
	variant dialogVariant
	props   DialogRootProps
	open    *reactive.Signal[bool]
}

func newDialogState(variant dialogVariant, props DialogRootProps) *dialogState {
	prefix := "qg-dialog"
	if variant == dialogVariantAlertDialog {
		prefix = "qg-alert-dialog"
	}
	return &dialogState{
		scope:   createComponentScope(prefix),
		variant: variant,
		props:   props,
		open:    reactive.NewSignal(props.DefaultOpen),
	}
}

func (state *dialogState) isOpen() bool {
	if state.props.Open != nil {
		return state.props.Open()
	}
	return state.open.Read()
}

func (state *dialogState) dismissOnEscape() bool {
	if state.props.DismissOnEscape != nil {
		return *state.props.DismissOnEscape
	}
	return true
}

func (state *dialogState) dismissOnBackdrop() bool {
	if state.props.DismissOnBackdrop != nil {
		return *state.props.DismissOnBackdrop
	}
	return state.variant != dialogVariantAlertDialog
}

func (state *dialogState) change(next bool, reason DialogOpenChangeReason, event *native.Event) {
	if state.props.Open == nil {
		state.open.Write(next)
	}
	if state.props.OnOpenChange != nil {
		state.props.OnOpenChange(next, DialogOpenChangeDetails{Reason: reason, Event: event})
	}
}

func (state *dialogState) complete(next bool, event *native.Event) {
	if state.props.OnOpenChangeComplete != nil {
		state.props.OnOpenChangeComplete(next, event)
	}
}

func (state *dialogState) applyTo(node *native.Node, part string) {
	setPart(node, part, state.scope, "")
	setString(node, protocol.Variant, string(state.variant))
	node.Bind(func() {
		setExplicitBool(node, protocol.Open, state.isOpen())
	})
}

var dialogContext = createPartContext[dialogState]()

func requireDialog(part string) *dialogState {
	return requireContext(dialogContext, part, "Dialog.Root")
}

func createDialogRoot(variant dialogVariant, props DialogRootProps) *native.Node {
	state := newDialogState(variant, props)
	return reactive.Provide(dialogContext, state, func() *native.Node {
		if props.Children == nil {
			return Fragment(nil)
		}
		return renderComponent(props.Children)
	})
}

func createDialogPart(part, name string, props PartProps) *native.Node {
	state := requireDialog(name)
	node := createViewPart(props)
	state.applyTo(node, part)
	return finishPart(node, props)
}

func createDialogTrigger(props PartProps) *native.Node {
	state := requireDialog("Dialog.Trigger")
	node := createButtonPart(props)
	state.applyTo(node, protocol.PartDialogTrigger)
	setListener(node, protocol.EventClick, forwardClick(props.OnClick, func(event *native.Event) {
		state.change(true, DialogTriggerPress, event)
	}))
	return finishPart(node, props)
}

func createDialogPortal(props PartProps) *native.Node {
	state := requireDialog("Dialog.Portal")
	node := createViewPart(props)
	state.applyTo(node, protocol.PartDialog)
	setMilliseconds(node, protocol.EnterDuration, state.props.EnterDuration)
	setMilliseconds(node, protocol.ExitDuration, state.props.ExitDuration)
	setListener(node, protocol.EventComponentChange, func(event *native.Event) {
		details := ComponentChangeFromEvent(event)
		if details == nil || details.OpenChangeComplete == nil {
			return
		}
		state.complete(*details.OpenChangeComplete, event)
	})
	return finishPart(node, props)
}

func createDialogPopup(props DialogPopupProps) *native.Node {
	state := requireDialog("Dialog.Popup")
	node := createViewPart(props.PartProps)
	state.applyTo(node, protocol.PartDialogPopup)
	setExplicitBool(node, protocol.DismissOnEscape, state.dismissOnEscape())
	setExplicitBool(node, protocol.DismissOnPointerOutside, state.dismissOnBackdrop())
	setListener(node, protocol.EventDismiss, func(event *native.Event) {
		if props.OnDismiss != nil {
			props.OnDismiss(event)
		}
		if !event.DefaultPrevented {
			state.change(false, DialogDismiss, event)
		}
	})
	return finishPart(node, props.PartProps)
}

func createDialogClose(props PartProps) *native.Node {
	state := requireDialog("Dialog.Close")
	node := createButtonPart(props)
	state.applyTo(node, protocol.PartDialogClose)
	setListener(node, protocol.EventClick, forwardClick(props.OnClick, func(event *native.Event) {
		state.change(false, DialogClosePress, event)
	}))
	return finishPart(node, props)
}

// Dialog is a controlled in-window modal dialog.
var Dialog = dialogAPI{}

type dialogAPI struct{}

func (dialogAPI) Root(props DialogRootProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogRoot(dialogVariantDialog, props)
}

func (dialogAPI) Trigger(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogTrigger(props)
}

func (dialogAPI) Portal(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogPortal(props)
}

func (dialogAPI) Viewport(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogPart(protocol.PartDialogViewport, "Dialog.Viewport", props)
}

func (dialogAPI) Backdrop(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogPart(protocol.PartDialogBackdrop, "Dialog.Backdrop", props)
}

func (dialogAPI) Popup(props DialogPopupProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogPopup(props)
}

func (dialogAPI) Title(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogPart(protocol.PartDialogTitle, "Dialog.Title", props)
}

func (dialogAPI) Description(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogPart(protocol.PartDialogDescription, "Dialog.Description", props)
}

func (dialogAPI) Close(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogClose(props)
}

// AlertDialog is a consequential dialog whose backdrop does not dismiss by default.
var AlertDialog = alertDialogAPI{}

type alertDialogAPI struct{}

func (alertDialogAPI) Root(props DialogRootProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)

	return createDialogRoot(dialogVariantAlertDialog, props)
}

func (alertDialogAPI) Trigger(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogTrigger(props)
}

func (alertDialogAPI) Portal(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogPortal(props)
}

func (alertDialogAPI) Viewport(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return Dialog.Viewport(props)
}

func (alertDialogAPI) Backdrop(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return Dialog.Backdrop(props)
}

func (alertDialogAPI) Popup(props DialogPopupProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogPopup(props)
}

func (alertDialogAPI) Title(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return Dialog.Title(props)
}

func (alertDialogAPI) Description(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return Dialog.Description(props)
}

func (alertDialogAPI) Close(props PartProps, children ...any) *native.Node {
	props.Children = withPartChildren(props.Children, children)
	return createDialogClose(props)
}
